### DISCUSSION TITLE
TDD: testing a sound or function call

### DISCUSSION THREAD

**Discussion by DanielGennat** [neuefische/questions#166] [Web] [answered]
Hi,
ich möchte einen Test schreiben, der überprüft, ob bei meinem Timer `onComplete` ein Sound abgespielt wird. Oder zumindest, ob die entsprechende Funktion bzw. if-Bedingung abgerufen wird.
Leider habe ich nach einer Stunde googlen noch keinen wirklichen Ansatz gefunden.
Kann mir da jemand helfen? :)

**Comment by MerleMak** [comment, marked as answer]
Hi Daniel,
Have a look at [this documentation](https://jestjs.io/docs/expect#tohavebeencalled), you will need to work with a mock function. To render your Timer component you will also need the React Testing Library. 
Maybe this already helps :)

**Reply by DanielGennat** [reply]
Vielen Dank, das ist die Methode, die ich vermisst habe :)
Ich habe das jetzt mal umgesetzt, aber es kommt mir noch nicht ganz richtig vor. Daher würde ich mich über einen Review freuen :)

[Hier](https://github.com/DanielGennat/capstone-project/pull/8) der Pullrequest.

index.test.js
```
import Home from './index.js';

function Home(callback, remainingTime) {
  if (remainingTime === 0) {
    callback(playNotificationSound(url));
  }
}

describe('Home', () => {
  test('timer has expired', () => {
    const playNotificationSound = jest.fn();
    Home(playNotificationSound, 0);
    expect(playNotificationSound).toHaveBeenCalled();
  });

  test('timer has not expired yet', () => {
    const playNotificationSound = jest.fn();
    Home(playNotificationSound, 1);
    expect(playNotificationSound).not.toHaveBeenCalled();
  });
});
```

index.js
```
export default function Home() {
  const renderTime = ({ remainingTime }) => {
    const minutes = Math.floor(remainingTime / 60);
    const seconds = remainingTime % 60;

    if (remainingTime === 0) {
      //sound notification goes here
    } else if (seconds < 60 && minutes < 1) {
      return (
        <div>
          <Time>{seconds}</Time>
        </div>
      );
    } else if (seconds < 10) {
      return (
```

**Reply by NikRoe** [reply]
Hey Daniel,

uns ist aufgefallen, dass die Soundnotification, die nach Ablauf des Timers abgespielt werden soll, noch gar nicht existiert, deswegen auch nicht getestet werden kann.
Zudem musst du bei NextJS, anders als bei React, `jest` manuell installieren und eine Konfiguationsdatei erstellen, die `jest.config.js`. Weitere Infos findest du [hier](https://nextjs.org/docs/testing#setting-up-jest-with-the-rust-compiler). NextJS nutzt mittlerweile den rust compiler.

Ich würde den Test zeitlich noch ein wenig nach hinten schieben, bis deine App über all die nötigen Funktionalitäten verfügt, die auch Teil des Tests sein müssen.
Falls du neben der Dokumentation von Jest noch ein Beispiel aus unserer Kurszeit ansehen möchtest, habe ich dir einmal den Link zu Lenes [Repo](https://github.com/macfrei/pokemon-shopping-list/pull/6/files#diff-305f6f4a056434ce304aa64ffdb03fd2b82f09d14ec75949839f0b075cc0e93d) rausgesucht, wo wir Testing gemacht haben.

Hab' ein schönes Wochenende

**Reply by DanielGennat** [reply]
... aber der Sinn von TDD ist doch, dass man erst den Test schreibt und danach programmiert, weil es ja eine Entwicklungsstrategie ist. Oder habe ich das falsch verstanden?

**Reply by DanielGennat** [reply]
Die Audio funktioniert jetzt. Wenn ich das Abspielen der Audio jetzt aber für das Testing als Funktion auslagern will, komme ich bei folgendem Fehler nicht weiter: "playNotificationSound ist not a function"

<img width="750" alt="Bildschirmfoto 2022-09-26 um 15 30 28" src="https://user-images.githubusercontent.com/109812125/192289608-bcfb86c3-d26b-4615-8f8e-c72673f5d7b2.png">

[Link zum Repo](https://github.com/DanielGennat/capstone-project/pull/8)

**Reply by ThomasKirst** [reply]
Die zwei äusseren Funktionen musst du innerhalb deiner Komponente schreiben:

```js
export default function Home() {
  const [audio, setAudio] = useState(null);

  useEffect(() => {
    setAudio(new Audio('/sounds/expired-notification.mp3'));
  }, []);

  function playNotificationSound() {
    audio.play();
  }
```

Die Signatur der `renderTime` würde ich folgendermassen anpassen:
```js
export function renderTime(remainingTime, playNotificationSound) {
```

In der Counter-Komponente musst du dann entsprechend den Call auf die `renderTime`-Funktion anpassen, um ihr alle nötigen Informationen bzw. die Funktion mitzugeben, die dann den Sound spielt:

```jsx
<CountdownCircleTimer
  isPlaying
  duration={3}
  colors={['#49F6EC', '#dfe057', '#ff6666', '#b80a24']}
  colorsTime={[40, 30, 10, 0]}
>
  {({ remainingTime }) =>
    renderTime(remainingTime, playNotificationSound)
  }
</CountdownCircleTimer>
```

**Reply by DanielGennat** [reply]
Jetzt funktioniert die Audio wieder, aber der Test nicht.

<img width="750" alt="Bildschirmfoto 2022-09-26 um 16 40 03" src="https://user-images.githubusercontent.com/109812125/192305255-34eb65f3-556c-49db-b4a8-19a79f98dd9a.png">

[Link zum Repo](https://github.com/DanielGennat/capstone-project/pull/8)

**Reply by DanielGennat** [reply]
final version:

<img width="750" alt="Bildschirmfoto 2022-09-26 um 22 39 19" src="https://user-images.githubusercontent.com/109812125/192375499-ae1823ff-0c0c-4e7d-996c-f5a98f357651.png">
